### ISSUE TITLE
Range parsing precedence

### ISSUE THREAD

**Issue by aljazerzen** [PRQL/prql#1837]
Currently, ranges support:

```
0..40
(0)..40
11..
..9
(-5..)
-10..
@2020-01-01..
@2020-01-01..@2021-01-01
```

There a few changes I'd propose to change.

## Only literals

With current grammar, `..` has a rule to only bind with literals:

```
take 1..@2023-02-14  # range from 1 to @2023-02-14
take 1..[]           # two arguments: range from 1 onward and an empty list
take 1..hello        # two arguments: range from 1 onward and an ident
```

Because literals are allowed to be enclosed in parenthesis, ranges are support nested expressions, but only if they contain literals only. 

Current parsing grammar thus contains two types of parenthesis: one for normal nesting and pipeline and the other for the ranges.

This is a weird rule and I propose we eliminate it, by using normal parenthesis for ranges too.

```
(1)..40     # ok
(-1)..40    # ok
(0-1)..40   # not ok currenty, I propose we allow this
(5 | abs | hello [false])..40   # not ok currenty, I propose we allow this
```
There is not a strong case that this would greatly improve some use case, I want to change it in name of consistency.

## Whitespace

In my recent push to remove effects of whitespace from PRQL, ranges proved to be the only breaking point.

```
take 1+1    # a single bin op
take 1+ 1   # a single bin op
take 1..1   # a single range
take 1.. 1  # two arguments: a range and an int
```

Whitespace is not required to bind the literal into range, but if the term is not a literal (i.e. a list) it will not be bound into range even if there is no whitespace.

I propose we remove this distinction and bind ranges the same way we bind all other binary operations.

## Result

```
take (10 - 5 + 2 * 2)..1
take [1, 10]..[5, 15]
take 1 .. 4
take 1 ..
take ( .. 4)  # parenthesis needed
```

**Comment by aljazerzen** [comment]
#1818 implements proposed behavior

**Comment by max-sixty** [comment]
Agree with "Only literals" point, a nice consistency win.

For "Whitespace":
- I do think the ergonomics of `take (..1)` are worse than `take ..1`. I'm not sure how I'd explain it without giving the whole explanation about parsing whitespace, which suggests we're too anchored in the parser relative to the language. 
- But it's also not the biggest deal, and I don't want to slow you down. 
- Do you have a view for why Rust allows for the `..1`? (I'm trying to think of other languages that I've thought of as whitespace-agnostic, this is the only case I can think of...).
- Is it even possible in Chumsky to generally ignore whitespace but not in a specific place? I can look later this week; I wonder whether it's possible to define the range as a unit that can't have whitespace in the same way that `->` or `>=` can't have whitespace (I'm presuming those can't have whitespace and we're only discussing the whitespace _between_ parsing items?)
  - Edit: i guess this is the lexer vs parser?

Other places where I had to care about whitespace when writing the pest parser:
- `s"foo"` as distinct from a function call on `s` — `s "foo"`
- `1.5` as distinct from two numbers, the first without any digits after the period `1. 5`
- A datetime like `@2021-12-31T12:00` as opposed to a date and named parameter `T12` with an arg `00`: `@2021-12-31 T12:00`
- Type definitions like `col<int>` as distinct from the comparison `col < int`?

Maybe I try and add some tests for these before we merge that branch...

**Comment by aljazerzen** [comment]
> Is it even possible in Chumsky to generally ignore whitespace but not in a specific place? Edit: i guess this is the lexer vs parser?

Yes, this is lexer vs parser. Lexer handles all whitespace related things, and so s-strings, numbers and dates can be parsed correctly.

I used to lex whitespace into `Token::Whitespace` that was then consumed by parser, but it was causing many problems and made parser definitions much more verbose.

So basically, information about the whitespace is not passed from lexer to parser. I just realized that I could convert `..` from being passed as other control token, to a special token that would know if it can bind to left and to right, depending on the whitespace:

```
Token::range { open_left: bool, open_right: bool }
```

---

Type definitions are gonna be tricky to implement too. 

---
